simde/x86/avx512/ror.h

```diff
@@ -0,0 +1,276 @@
+#if !defined(SIMDE_X86_AVX512_ROR_H)
+#define SIMDE_X86_AVX512_ROR_H
+
+#include "types.h"
+#include "mov.h"
+#include "or.h"
+#include "srli.h"
+#include "slli.h"
+#include "../avx2.h"
+
+HEDLEY_DIAGNOSTIC_PUSH
+SIMDE_DISABLE_UNWANTED_DIAGNOSTICS
+SIMDE_BEGIN_DECLS_
+
+#if defined(SIMDE_X86_AVX512F_NATIVE) && defined(SIMDE_X86_AVX512VL_NATIVE)
+  #define simde_mm_ror_epi32(a, imm8) _mm_ror_epi32(a, imm8)
+#else
+  SIMDE_FUNCTION_ATTRIBUTES
+  simde__m128i
+  simde_mm_ror_epi32_internal_(simde__m128i a, simde__m128i result, int imm8)
+      SIMDE_REQUIRE_CONSTANT_RANGE (imm8, 0, 255) {
+    simde__m128i retval;
+
+    if(imm8 & 31)
+      retval = result;
+    else
+      retval = a;
+
+    return retval;
+  }
+  #define simde_mm_ror_epi32(a, imm8) simde_mm_ror_epi32_internal_(a, simde_mm_or_si128(simde_mm_srli_epi32(a, (imm8 & 31)), simde_mm_slli_epi32(a, (32 - (imm8 & 31)))), imm8)
```

Comment:
This won't work.  You're using `a` multiple times in the macro, which can easily cause horrible problems.  An easy to understand example would be something like

```c
r = simde_mm_ror_epi32((value = simde_mm_add_epi32(value, other_value)), 4);
```

However, it can be a lot more subtle.  For example, what if instead of `simde_mm_add_epi32` they pass the result of a function which does some expensive computation?  Or if you have a pass-by-reference parameter it's really easy to mess something up, especially in C++ where the function call looks exactly the same.

It's okay to do this sort of thing with the `imm8` parameter because they are required to be an integer constant expression, but for "normal" parameters it's a big problem.  And, since that's an issue on the *caller* side, there isn't really anything we can do to prevent it in SIMDe.

For most compilers you can work around the issue using a [statement expression](https://gcc.gnu.org/onlinedocs/gcc/Statement-Exprs.html); we even have a macro (`SIMDE_STATEMENT_EXPR_`) which you can check to see if statement expressions are supported.  That would look something like:

```c
#if defined(SIMDE_STATEMENT_EXPR_)
  #define simde_mm_ror_epi32(a, imm8) \
    SIMDE_STATEMENT_EXPR_(({ \
      simde_mm_ror_epi32_a_ = a; \
      simde_mm_ror_epi32_internal_( \
        simde_mm_ror_epi32_a_, \
        simde_mm_or_si128( \
          simde_mm_srli_epi32(simde_mm_ror_epi32_a_, \
            (imm8 & 31) \
          ), \
          simde_mm_slli_epi32( \
            simde_mm_ror_epi32_a_, \
            (32 - (imm8 & 31)) \
          ) \
        ), \
      imm8) \
    }))
#endif
```

However, this doesn't work on MSVC.

This might mean that you can't use `simde_mm_srli_epi32` and `simde_mm_slli_epi32`, but that's probably okay in this case.  This is a pretty simple function which I would expect most compilers to be able to autovectorize well, especially those which support GCC-style vector extensions (where you can just do `(a_.u32 >> count) | (a_.u32 << count)`, though of course that *also* doesn't include MSVC).

Reply:
You should also be wrapping variables used in the macro in parenthesis.  For a discussion of why this is important, see https://stackoverflow.com/questions/7186504/c-macros-and-use-of-arguments-in-parentheses, but it comes down to the same issue; you're not passing the result of the expression, you're passing the entire expression (which may have side-effects).

Reply:
Thanks, I will fix this and keep that in mind